### PR TITLE
[9.x] Makes blade components blazing fast ⛽️

### PR DESCRIPTION
**Pull request by nunomaduro**
## Technical Details 

This pull request addresses the performance of blade view components, and it mainly performs:

1. Components are only resolved from the IOC container if necessary. This is done by checking if the component `$data` matches exactly the `Component::__construct()` method parameters names, and if not, we proxy for the container. 

2. The computation of the `Component::render`, when using inline components, is only computed once. So, assuming that the return is `components.button`, we only check if `components.button` is a view once, and following calls will use this cached result.

3. We provide a new method, called `Component::view`, that should be documented instead of having people using the simple `view` helper.

4. Events on views, related to creators and composers, are only fired if there is actually creators or composers for the given view.

5. Checking if views are expired, only happens once for the given view, during the lifecycle of the Application instance. 

## How To Test This Pull Request

If you would like to test this pull request locally, simply run the command above.

```
composer require laravel/framework:"dev-feat/view-components-performance as 9.34" --with-all-dependencies
php artisan clear
php artisan view:clear
```

After, run your test suite, navigate in your application, and see if things are working as normal.

## Results
**Anonymous components — up to 52% faster:** 
```
@for ($number = 0; $number < $numberOfComponents; $number++)
    <x-button: :$number /> 
@endfor

Before:
- 100 components: 7.203ms
- 1000 components: 67.764ms
- 10000 components: 666.120ms
After:
- 100 components: 3.649ms
- 1000 components: 34.204ms
- 10000 components: 326.806ms
```

**Class View components - up to 39% faster:** 
```
@for ($number = 0; $number < $numberOfComponents; $number++)
    <x-button: :$number /> 
@endfor

class Button extends Component
{
    function __construct(public $number) {}

    function render() {
        return 'components.button';
    }
}

Before:
- 100 components: 6.946ms
- 1000 components: 69.657ms
- 10000 components: 670.511ms
After:
- 100 components: 4.441ms
- 1000 components: 42.266ms
- 10000 components: 413.125ms
```

**Class Inline components — up to 92% faster:** 
```
@for ($number = 0; $number < $numberOfComponents; $number++)
    <x-button: :$number /> 
@endfor

class Button extends Component
{
    function __construct(public $number) {}

    function render() {
        return '{{ $number }}';
    }
}

Before:
- 100 components: 66.426ms
- 1000 components: 560.284ms" // 0.5 seconds ..
- 10000 components: 5,080.315ms" // 5 seconds ..
After:
- 100 components: 4.499ms
- 1000 components: 41.834ms
- 10000 components: 413.493ms
```

**Dynamic components ( that renders anonymous components ) — up to 72% faster:** 
```
@for ($number = 0; $number < $numberOfComponents; $number++)
    <x-dynamic-component :component="'button'" :$number />
@endfor

Before:
- 100 components: 239.325ms
- 1000 components: 1,033.402ms
- 10000 components: 6,994.827ms

After:
- 100 components: 185.298ms
- 1000 components: 416.685ms
- 10000 components: 1,994.571ms
```

**Dynamic components ( that renders inline class components ) — up to 83% faster:** 
```
@for ($number = 0; $number < $numberOfComponents; $number++)
    <x-dynamic-component :component="'button'" :$number />
@endfor

class Button extends Component
{
    function __construct(public $number) {}

    function render() {
        return '{{ $number }}';
    }
}

Before:
- 100 components: 287.660ms
- 1000 components: 1,331.259ms
- 10000 components: 12,439.949ms

After:
- 100 components: 69.568ms
- 1000 components: 227.673ms
- 10000 components: 2,216.089ms
```

### CODE REVIEW

**Comment by JayBizzle** [comment]
Just tested this again against the issue we were having https://github.com/blade-ui-kit/blade-icons/issues/190

Based on the repo in that issue we are seeing the following improvement when rendering 10,000 icons with this PR in place (M1 Mac Studio)

```
730.799ms   // component without params
1,233.669ms // component with param
```

After this PR applied

```
272.211ms  // component without params
273.206ms  // component with param
```

👍

**Comment by garygreen** [comment]
This PR is quite large. Could we have a better understanding of what parts of this PR contribute to the speed increase?

It looks like there are 3 main areas that are being targeted here to improve speed:

- [blade compiled cache check](https://github.com/laravel/framework/pull/44487/files#diff-e150e8bf5115ebcf8c4a5e2bb44141867e37f3a6e69fe7dbc96c2acc1caae670R61)
- [composers/creators skipping from being called](https://github.com/laravel/framework/pull/44487/files#diff-6cf51d9470b4d74d122ba56d1215e9402c7f5fe505d42a5e455755ef1b0b2205R35)
- [the way the component is resolved](https://github.com/laravel/framework/pull/44487/files#diff-1005052c893d685dfa5b9c5c7431e2a880260f9a16cba75e16689ec33549755cR397)

What parts of those contribute most to the speed increase? Are they all necessary, or could some of them be micro optimizations?

File: src/Illuminate/View/Component.php [hunk 1]
```diff
@@ -292,4 +325,121 @@ public function shouldRender()
...
+
+    /**
+     * Get the evaluated view contents for the given view.
+     *
+     * @param  string|null  $view
+     * @param  \Illuminate\Contracts\Support\Arrayable|array  $data
+     * @param  array  $mergeData
+     * @return \Illuminate\Contracts\View\View
+     */
+    public function view($view, $data = [], $mergeData = [])
```
Comment (nunomaduro) [inline src/Illuminate/View/Component.php:350]: Note here: seems that, declaring this method, conflicts with the https://filamentphp.com's component class. So, shipping this new method, makes projects using older versions of https://filamentphp.com to break.

So, the filament team will have to adjust to this change, or we can ship this only in Laravel X.

Comment (nunomaduro) [reply src/Illuminate/View/Component.php:350]: This new method is not required to have speed improvements, it's like a 7% speed improvement at scale.

Comment (danharrin) [reply src/Illuminate/View/Component.php:350]: If this is merged we will merge filamentphp/filament#4465 to mitigate the problem

**Comment by ErikBernskiold** [comment]
Love this! I just tried this on a project which relies heavily on data tables where tables, tr, td and th are all anonymous components. The experience has been sluggish on render. While I couldn't do the proper benchmarking at the moment, the difference is night and day and _very_ noticeable. Blazing fast!

**Comment by browner12** [comment]
Just an FYI, this breaks the "View" tab of Laravel Debugbar.  I'll go make an issue over there as well.

Screen Shot 2022-10-18 at 11 06 06 AM

File: src/Illuminate/View/Engines/CompilerEngine.php [hunk 2]
```diff
@@ -51,10 +58,12 @@ public function get($path, array $data = [])
         // If this given view has expired, which means it has simply been edited since
         // it was last compiled, we will re-compile the views so we can evaluate a
         // fresh copy of the view. We'll pass the compiler the path of the view.
-        if ($this->compiler->isExpired($path)) {
+        if (! isset($this->compiledOrNotExpired[$path]) && $this->compiler->isExpired($path)) {
```
Comment (erikn69) [inline src/Illuminate/View/Engines/CompilerEngine.php:61]: @nunomaduro this is giving me a Exception https://github.com/spatie/laravel-permission/actions/runs/3276093280/jobs/5391731843 after `9.36.0`
maybe must be
```diff
-  if (! isset($this->compiledOrNotExpired[$path]) && $this->compiler->isExpired($path)) {
+  if (! isset($this->compiledOrNotExpired[$path]) || $this->compiler->isExpired($path)) {
```
Or i'm doing something wrong, nothing changed on that package

btw, thanks for the great work

Comment (nunomaduro) [reply src/Illuminate/View/Engines/CompilerEngine.php:61]: Fixed: https://github.com/laravel/framework/pull/44643.

**Comment by jhoanborges** [comment]
BRAVO!!! @nunomaduro

**Comment by hoRacy** [comment]
A beautiful change and colossal speed difference. Thank you very much for your contribution to the project! <3 @nunomaduro

**Comment by Marcel-Sass** [comment]
When using:
`DbView::make($mailConfig, $data, [], 'content') ->render();`
it produces following error: 
`{
    "class": "TypeError",
    "message": "Illegal offset type in isset or empty",
    "code": 0,
    "file": "/Users/sassm/Documents/PHPStormProjects/bw-shop/vendor/laravel/framework/src/Illuminate/View/Engines/CompilerEngine.php:61",
    "trace": [
        "/Users/sassm/Documents/PHPStormProjects/bw-shop/vendor/laravel/framework/src/Illuminate/View/View.php:139",
        "/Users/sassm/Documents/PHPStormProjects/bw-shop/vendor/flynsarmy/db-blade-compiler/src/Flynsarmy/DbBladeCompiler/DbView.php:113",
        "/Users/sassm/Documents/PHPStormProjects/bw-shop/vendor/flynsarmy/db-blade-compiler/src/Flynsarmy/DbBladeCompiler/DbView.php:98",
        "/Users/sassm/Documents/PHPStormProjects/bw-shop/vendor/flynsarmy/db-blade-compiler/src/Flynsarmy/DbBladeCompiler/DbView.php:70",
        "/Users/sassm/Documents/PHPStormProjects/bw-shop/app/Helper/Mail/MailHelper.php:54",`

**Pull request merged**